go/batch-submitter/drivers/sequencer/encoding.go

```diff
@@ -168,6 +239,34 @@ func (p *AppendSequencerBatchParams) Read(r io.Reader) error {
 		p.Contexts = append(p.Contexts, batchContext)
 	}
 
+	p.Type = BatchTypeLegacy
+
+	// Handle backwards compatible batch types
+	if len(p.Contexts) > 0 && p.Contexts[0].Timestamp == 0 {
+		switch p.Contexts[0].BlockNumber {
+		case 0:
+			// zlib compressed transaction data
+			p.Type = BatchTypeZlib
+			// remove the first dummy context
+			p.Contexts = p.Contexts[1:]
+			numContexts--
+
+			zr, err := zlib.NewReader(r)
+			if err == io.EOF || err == io.ErrUnexpectedEOF {
+				return nil
+			} else if err != nil {
+				return err
+			}
+			defer zr.Close()
+
+			buf := new(bytes.Buffer)
```

Comment:
The TX deserialization code below uses an `io.Reader`, which `zlib.NewReader` implements. This means that you could just set `r = zr` and avoid buffering the entire transaction in memory. Curious if the choice to buffer was a deliberate one.

Reply:
I will double check this is still the case because I did refactor the code a bit, but I this because the reader index was at the end of the underlying slice, which caused an eof error later in the codepath

Reply:
The following diff results in the tests breaking:

```diff
diff --git a/go/batch-submitter/drivers/sequencer/encoding.go b/go/batch-submitter/drivers/sequencer/encoding.go
index c428b564..76aa441c 100644
--- a/go/batch-submitter/drivers/sequencer/encoding.go
+++ b/go/batch-submitter/drivers/sequencer/encoding.go
@@ -259,11 +259,14 @@ func (p *AppendSequencerBatchParams) Read(r io.Reader) error {
 			}
 			defer zr.Close()
 
-			buf := new(bytes.Buffer)
-			if _, err := io.Copy(buf, zr); err != nil {
-				return err
-			}
-			r = bytes.NewReader(buf.Bytes())
+			/*
+				buf := new(bytes.Buffer)
+				if _, err := io.Copy(buf, zr); err != nil {
+					return err
+				}
+				r = bytes.NewReader(buf.Bytes())
+			*/
+			r = zr
 		}
 	}
```

I believe it is due to the reading of too much data here:

https://github.com/ethereum-optimism/optimism/blob/43f16dd06b7173a6b8a6703fec984df8b2ea91c2/go/batch-submitter/drivers/sequencer/encoding.go#L210-L214

The 8 bytes have been compressed and we have no way to know how many bytes to read without decompressing the entire payload first (unless we break the `io.Reader` abstraction and use knowledge of the underlying compression algo). I think this is more simple for now.

One change that I can make is this:

```diff
diff --git a/go/batch-submitter/drivers/sequencer/encoding.go b/go/batch-submitter/drivers/sequencer/encoding.go
index c428b564..48c15dd6 100644
--- a/go/batch-submitter/drivers/sequencer/encoding.go
+++ b/go/batch-submitter/drivers/sequencer/encoding.go
@@ -263,7 +263,7 @@ func (p *AppendSequencerBatchParams) Read(r io.Reader) error {
 			if _, err := io.Copy(buf, zr); err != nil {
 				return err
 			}
-			r = bytes.NewReader(buf.Bytes())
+			r = buf
 		}
 	}
 ```

Reply:
hmm something seems weird, it should still have those bytes buffered in the zlib writer. not sure what's going on tho 🤔

Reply:
I still think that the problem here is the explicit usage of telling the reader to read 8 bytes, we want to read the next item which was 8 bytes before compression but now that its compressed we end up reading too much which offsets the reads for the rest of the datastructure